koding/koding
# kloud/provider: Add SoftLayer Support

Adds SoftLayer support as a kloud provider, and necessary frontend changes.

## Description
Following the example here [https://github.com/koding/kloud-provider-example](https://github.com/koding/kloud-provider-example) and the PRs for [DigitalOcean](https://github.com/koding/koding/pull/9187) and [Google](https://github.com/koding/koding/pull/9185) support, this PR include the necessary changes to enable SoftLayer provider support for kloud.

Part of this PR includes updates to the vendor Terraform source in this repository, and essentially consists of getting the SoftLayer provider directory from Terraform's master branch and dropping it in place here. The current SoftLayer support that Terraform has is basically provisioning virtual guests (or vms) and managing ssh keys.

This PR also includes most (maybe all?) of the UI/frontend changes necessary for a user to use Koding with SoftLayer.

## Motivation and Context
Implement a solution to #9073.

## How Has This Been Tested?
- Unit tests to validate basic functionality for `stack` `machine` and `schema` definitions
- Integration tests that rely on environment variables to actually call SoftLayer apis where necessary in validating `stack` and `machine` definitions
- I found mentions on Slack channel about the `downtowork` koding development deployment, so I created a development environment [http://downtowork-acbodine.koding.team:8090](http://downtowork-acbodine.koding.team:8090). In order to test my changes with the whole system I simply pulled my changes into this enviroment and rebuilt/re-ran the necessary Docker containers. You can create a team there, or use the one I have been developing with [http://dragons.downtowork-acbodine.koding.team:8090](http://dragons.downtowork-acbodine.koding.team:8090) and you can leverage the SoftLayer stack I have in that team `Voltorb SoftLayer Stack`
- The frontend/user testing that I did has been manual so far in order to verify that I could
  1. Verify Credentials user enters
  2. Save SoftLayer stack and generate Bootstrap resources, verified in SoftLayer UI
  3. Initialize SoftLayer virtual guest, and verified connectivity back to Koding client/UI
  4. Tested Koding UI displaying accurate virtual guest state info, as well as stop/start actions

## Types of changes

## Checklist:
